primitives/core/src/network.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020-2023 Litentry Technologies GmbH.
+// This file is part of Litentry.
+//
+// Litentry is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{traits::ConstU32, BoundedVec};
+use sp_std::{hash::Hash, vec::Vec};
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
+
+pub const MAX_WEB3NETWORK_LEN: u32 = 128;
+pub type BoundedWeb3Network = BoundedVec<Web3Network, ConstU32<MAX_WEB3NETWORK_LEN>>;
+
+/// supported web3 networks
+/// use a flattened style to avoid overly nested structure like:
+/// {
+///   Substrate(SubstrateNetwork),
+///   Evm(EvmNetwork),
+/// }
+/// TODO: theoretically this should the the union of the supported networks of all data providers
+#[derive(
+	Encode,
+	Decode,
+	Copy,
+	Clone,
+	Debug,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Hash,
+	TypeInfo,
+	MaxEncodedLen,
+	EnumIter,
+)]
+pub enum Web3Network {
+	// substrate
```

Comment:
It was something like that before the change. One more structure layer in between. Either way has its stand. I think here Kai wants to decrease the layers.

Reply:
Hi @kziemianek, yes I intentionally did so after pondering it 🥲 see the comment just above this struct:
```
/// supported web3 networks
/// use a flattened style to avoid overly nested structure like:
/// {
///   Substrate(SubstrateNetwork),
///   Evm(EvmNetwork),
/// }
```

The nested style would make it more complicated when constructing such type (also from F/E side). The value retrieval and comparison/branching typically need more effort (`if let`/`match`) in code which lowers the readability.  E.g. 
```
Web3Network::Litentry
vs
WebNetwork::Substrate(SubstrateNetwork::Litentry)
```
Both "Substrate" are kind of redundant too.
You can have extra `From` or `TryFrom` implementation but that's not as straightforward either.

If some of the enum types (e.g. `Ethereum`) need further subdivision, that's even more layers. `is_substrate` and `is_evm` would take a one-time maintenance effort so I'm still fine with it.

Adding new values is a fair point, but either way it would need a runtime upgrade to take effect => it will have a new spec_version number => it signals the client it's a breaking change => the client needs to refresh/renew the metadata eventually. That said I'll keep it categorised when adding new values.

So in general I consider the flattened structure a bit more advantageous but I'm open to suggestions.

Reply:
_Adding new values is a fair point, but either way it would need a runtime upgrade to take effect => it will have a new spec_version number => it signals the client it's a breaking change => the client needs to refresh/renew the metadata eventually. That said I'll keep it categorised when adding new values._

This might not be true for clients that are using DI only.

This scaries me a bit more, because if you didn't notice the change, you might be sending old values and everything will work fine, except the results will be not the ones you expected.

Reply:
Staing with this structure but not forcing new variants to be added "in between" should avoid such problems.

Reply:
That's akin to the substrate extrinsic index - where adding new extrinsic is backwards compatible, but if you change the order that's a breaking change and you have to bump the `spec_version`/`transaction_version`.

I'd argue even if you split into `evm` and `substrate` groups you can still fall into this problem once you remove/change the existing fields. The same applies to other enums too (remember the famous `trusted_call`).

I'm not sure how IDHub maintains the generated types, but maybe we can minimise the risk by having a proper version and metadata update mechanism, like utilise the runtime version in sidechain too - once the client sees a different runtime version, it should realise something is broken and they should re-fetch the metadata and re-generate the types.

Reply:
I also agree that we should look for structures that keep breaking changes to a minimum or are at least backward compatible in the future without translating the problem to clients.

put another way, as a client, if you give to pick between ugly and nested structures that require mininum to non maintenance and nice structures that require me to watch every single thing you change, I'd very much prefer the former.

Now, I leave it to you folks to find such a structure and keep a good balance between both trade-offs 👍